// @abaplint/runtime/build/src/builtin/sy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sy = void 0;
const types_1 = require("../types");
exports.sy = new types_1.Structure({
    abcde: new types_1.Character(26).set("ABCDEFGHIJKLMNOPQRSTUVWXYZ"),
    datlo: new types_1.Date(),
    datum: new types_1.Date(),
    dbcnt: new types_1.Integer(),
    fdpos: new types_1.Integer(),
    host: new types_1.Character(32).set("localhost"),
    index: new types_1.Integer(),
    langu: new types_1.Character(1).set("E"),
    mandt: new types_1.Character(3).set("123"),
    msgid: new types_1.Character(20),
    msgno: new types_1.Numc({ length: 3 }),
    msgty: new types_1.Character(1),
    msgv1: new types_1.Character(50),
    msgv2: new types_1.Character(50),
    msgv3: new types_1.Character(50),
    msgv4: new types_1.Character(50),
    subrc: new types_1.Integer(),
    sysid: new types_1.Character(3).set("ABC"),
    tabix: new types_1.Integer(),
    tfill: new types_1.Integer(),
    timlo: new types_1.Time(),
    tzone: new types_1.Integer(), // 0 = UTC
    uname: new types_1.Character(12).set("USERNAME"),
    uzeit: new types_1.Time(),
    dbsys: new types_1.Character(10),
    batch: new types_1.Character(1),
});
